### DISCUSSION TITLE
Accessing $(GitVersion_*) variables from MSBuild

### DISCUSSION THREAD

**Discussion by professor-k** [GitTools/GitVersion#3410] [Q&A]
It's probably stupid question, but I still need to ask it. I can't seem to access GitVersion_* msbuild variables from AfterBuild events. I was able to reduce it with just few trivial lines:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>net6.0</TargetFrameworks>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="GitVersion.MsBuild" Version="5.12.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>
  <Target Name="AfterBuild" AfterTargets="Build">
    <Message Text="Build number '$(GitVersion_FullSemVer)'" Importance="High" />
  </Target>
</Project>
```
It's doesn't seem that the issue depends on repository or GitVersion.yml (or absence of it).  GitVersion intermediate files in obj folder are produced, resulting dll has respective attributes set, everything seems okay. When build with diag logging, it's clear that variables are set too. However when it gets to AfterBuild event, they are either empty or unset.

I must be missing something obvious here, because same constructs work in some other of my projects (way more complicated, with few levels of Directory.build.props and complex dependencies).

Any help would be appreciated.

**Comment by professor-k** [comment]
Okay, I figured the difference between working and not working cases. Working one looks like this: `<Target Name="PrintVersion" AfterTargets="GetVersion">`
What confused the hell out of me, was that it won't get called at all if you declare it like this: `<Target Name="AfterBuild" AfterTargets="GetVersion">`. There must be some magic arount this particular target name. Still not sure why it doesn't work with actual `AfterTargets="Build"`.

p.s. And `<Target Name="anythingbutafterbuild" AfterTargets="Build">` will get called twice, once with value inside, and once without. 🤯

**Comment by arturcic** [comment]
The reason it does not work after "Build" is that the "GetVersion" runs after "Build", meaning the said variables are set by the "GitVersion" target and not "Build"

**Reply by professor-k** [reply]
Is that right? As far as I see (looking into 5.12.0 nuget package right now), GetVersion is actually called before bunch of targets including BeforeCompile (which makes sense, as it is part of the group of targets that generate *.g.cs files that have to be included in compilation).
